### PR TITLE
win32: Cleanup fs code for WinSCP support

### PR DESCRIPTION
**Pull request by remittor**
Clenup WIN32 wrapper code for support SFTP via WinSCP utility

### CODE REVIEW

**Comment by remittor** [comment]
openssh-winscp

**Comment by remittor** [comment]
openssh-longpath

File: contrib/win32/win32compat/fileio.c [hunk 1]
```diff
@@ -256,12 +258,26 @@ struct w32_io*
...
-            debug("utf8_to_utf16 failed - ERROR:%d", GetLastError());
-            return NULL;
-    }
-        
+	if ((path_utf16 = utf8_to_utf16(path_utf8)) == NULL) {
+		errno = ENOMEM;
+		debug("utf8_to_utf16 failed - ERROR:%d", GetLastError());
+		return NULL;
+	}
+	if (wcslen(path_utf16) >= MAX_PATH - 2) {
```
Comment (bagajjal) [inline contrib/win32/win32compat/fileio.c:266]: plz sync the latest code (v0.0.7.0) from latestw_all branch and use PATH_MAX instead of MAX_PATH.

File: contrib/win32/win32compat/fileio.c [hunk 2]
```diff
@@ -256,12 +258,26 @@ struct w32_io*
...
+			return NULL;  /* support only full path */
+		}
+		free(path_utf16);
+		if ((path_utf16 = utf8_to_wchar("\\\\?\\%s", path_utf8)) == NULL) {
+			errno = ENOMEM;
+			debug("utf8_to_wchar failed - ERROR:%d", GetLastError());
+			return NULL;
+		}
+	}
+	convertToBackslashW(path_utf16);
```
Comment (bagajjal) [inline contrib/win32/win32compat/fileio.c:279]: I think, this is not required as the winAPIs are able to recognize the forward slash paths as well.

File: contrib/win32/win32compat/fileio.c [hunk 3]
```diff
@@ -564,15 +580,28 @@ fileio_fstat(struct w32_io* pio, struct _stat64 *buf) {
 
 int
 fileio_stat(const char *path, struct _stat64 *buf) {
-    wchar_t wpath[MAX_PATH];
+    int ret = -1;
+    wchar_t wpath[8];
```
Comment (bagajjal) [inline contrib/win32/win32compat/fileio.c:584]: please revert the changes in this function as it is fixed in the right way as it is not specific to "ls command".
The latest version (V0.0.7.0), latestw_all  branch has these changes.

FYI, The latest version fixes the issue of "ls c:".

File: contrib/win32/win32compat/fileio.c [hunk 4]
```diff
@@ -564,15 +580,28 @@ fileio_fstat(struct w32_io* pio, struct _stat64 *buf) {
...
     wchar_t* wtmp = NULL;
 
+    if (path && strcmp(path, "/") == 0) {
```
Comment (bagajjal) [inline contrib/win32/win32compat/fileio.c:587]: I think this if loop is still required to support the WINSCP changes you have made.

File: contrib/win32/win32compat/inc/sys/param.h [hunk 5]
```diff
@@ -6,5 +6,10 @@ typedef unsigned int gid_t;
 typedef long long  off_t;
 typedef unsigned int dev_t;
 
+#undef NAME_MAX
+#define NAME_MAX   255
+
+#undef PATH_MAX
```
Comment (bagajjal) [inline contrib/win32/win32compat/inc/sys/param.h:12]: revert back this change as the PATH_MAX is now pointing to MAX_PATH.

PATH_MAX can't be 4096 as the current windows APIs accepts MAX_PATH length only. If we need to support path lengths more than MAX_PATH then the windows APIs have to be changed.
FYI, https://msdn.microsoft.com/en-us/library/windows/desktop/aa365247(v=vs.85).aspx will explain it clearly about the paths more than MAX_PATH.

File: contrib/win32/win32compat/fileio.c [hunk 6]
```diff
@@ -256,12 +258,26 @@ struct w32_io*
...
+		return NULL;
+	}
+	if (wcslen(path_utf16) >= MAX_PATH - 2) {
+		if (path_utf16[1] != ':') {
+			free(path_utf16);
+			errno = ENODEV;
+			return NULL;  /* support only full path */
+		}
+		free(path_utf16);
+		if ((path_utf16 = utf8_to_wchar("\\\\?\\%s", path_utf8)) == NULL) {
```
Comment (bagajjal) [inline contrib/win32/win32compat/fileio.c:273]: why is this required? the current code doesn't support the file name > MAX_PATH.. to make this happen, there are many places where we need to change the code to use the appropriate APIs (r: CopyFileW, CreateFileW, CreateFile2, DeleteFileW, FindFirstFileW, FindFirstFileExW, SetFileAttributesW, GetFullPathNameW, etc).

File: contrib/win32/win32compat/misc.c [hunk 7]
```diff
@@ -176,15 +176,47 @@ w32_fopen_utf8(const char *path, const char *mode) {
...
+		if (!utf8)
+			return NULL;
+		if (*utf8 == 0)
+			return wcsdup(L"");
+	}
+
+	if ((needed = MultiByteToWideChar(CP_UTF8, 0, utf8, -1, NULL, 0)) == 0)
+		goto exit;
+
+	if ((utf16 = malloc(needed * sizeof(wchar_t))) == NULL)
```
Comment (bagajjal) [inline contrib/win32/win32compat/misc.c:204]: combine this with the above if loop by using "||" to make the code simple.

File: contrib/win32/win32compat/misc.c [hunk 8]
```diff
@@ -549,13 +581,27 @@ w32_chdir(const char *dirname_utf8) {
 
 char *
 w32_getcwd(char *buffer, int maxlen) {
-	wchar_t wdirname[MAX_PATH];
+	wchar_t * wdir;
 	char* putf8 = NULL;
 
-	_wgetcwd(&wdirname[0], MAX_PATH);
+	if (buffer)
+		*buffer = 0;
```
Comment (bagajjal) [inline contrib/win32/win32compat/misc.c:588]: instead, use memcpy to initialize the whole buffer

File: contrib/win32/win32compat/misc.c [hunk 9]
```diff
@@ -617,23 +672,72 @@ convertToForwardslash(char *str) {
 * This method will resolves references to /./, /../ and extra '/' characters in the null-terminated string named by
 *  path to produce a canonicalized absolute pathname.
 */
-char *
-realpath(const char *path, char resolved[MAX_PATH]) {
-	char tempPath[MAX_PATH];
+char * realpath(const char *path, char *resolved) {
```
Comment (bagajjal) [inline contrib/win32/win32compat/misc.c:675]: Please sync the latest code and submit the new review as there are conflicting changes in this function.

File: contrib/win32/win32compat/misc.c [hunk 10]
```diff
@@ -617,23 +672,72 @@ convertToForwardslash(char *str) {
...
+		return NULL;
+
+	if (strcmp(path, "/") == 0 || strcmp(path, "/..") == 0 ||
+	    (path[0] == '/' && isalpha(path[1]) && strcmp(path + 2, ":/..") == 0)) {
+		if (!resolved)
+			return strdup("/");
+		strncpy(resolved, "/", PATH_MAX);
+		return resolved;
+	}
+	if (path[0] == '/' && isalpha(path[1]) && strcmp(path + 2, ":") == 0) {
```
Comment (bagajjal) [inline contrib/win32/win32compat/misc.c:694]: this if loop can be removed and the below code should be able to handle this.

File: contrib/win32/win32compat/misc.c [hunk 11]
```diff
@@ -617,23 +672,72 @@ convertToForwardslash(char *str) {
...
+
+	wpath = utf8_to_utf16(path);
+	if (wpath == NULL)
+		fatal("failed to convert input arguments");
 		
-	if ( (strlen(path) >= 2) && (path[0] == '/') && (path[2] == ':') )
-		strncpy(resolved, path + 1, strlen(path)); // skip the first '/'
-	else
-		strncpy(resolved, path, strlen(path) + 1);
+	wtmp = _wfullpath(NULL, wpath, 0);
```
Comment (bagajjal) [inline contrib/win32/win32compat/misc.c:708]: this code is not reqiured.. _fullpath() is sufficient as it just expands the ".", ".." and returns the full path.

File: contrib/win32/win32compat/misc.c [hunk 12]
```diff
@@ -772,7 +876,7 @@ int statvfs(const char *path, struct statvfs *buf) {
 		buf->f_favail = -1;
 		buf->f_fsid = 0;
 		buf->f_flag = 0;
-		buf->f_namemax = MAX_PATH - 1;
+		buf->f_namemax = NAME_MAX;
```
Comment (bagajjal) [inline contrib/win32/win32compat/misc.c:879]: not sure why it can't be PATH_MAX (which is MAX_PATH)?

File: contrib/win32/win32compat/win32_dirent.c [hunk 13]
```diff
@@ -11,33 +11,92 @@
...
+
 /* Open a directory stream on NAME.
    Return a DIR stream on the directory, or NULL if it could not be opened.  */
 DIR * opendir(const char *name)
 {
 	struct _wfinddata_t c_file;
 	intptr_t hFile;
 	DIR *pdir;
-	wchar_t searchstr[MAX_PATH];
+	wchar_t searchstr[PATH_MAX + 16];
```
Comment (bagajjal) [inline contrib/win32/win32compat/win32_dirent.c:74]: why 16? plz add comments.

File: contrib/win32/win32compat/win32_dirent.c [hunk 14]
```diff
@@ -62,15 +121,64 @@ DIR * opendir(const char *name)
    Return 0 if successful, -1 if not.  */
 int closedir(DIR *dirp)
 {
-    if ( dirp && (dirp->hFile) ) {
+    if (!dirp)
+        return 0;
+    if (dirp->hFile)
         _findclose( dirp->hFile );
```
Comment (bagajjal) [inline contrib/win32/win32compat/win32_dirent.c:127]: set 
dirp->hFile = 0

**Review by bagajjal — CHANGES_REQUESTED** [review]

File: contrib/win32/win32compat/win32_dirent.c (hunk 13 above)
Comment (remittor) [reply contrib/win32/win32compat/win32_dirent.c:74]: See below code:
`swprintf_s(searchstr, _ARRAYSIZE(searchstr) - 1, L"\\\\?\\%s\\*.*", wname);`

Comment (bagajjal) [reply contrib/win32/win32compat/win32_dirent.c:74]: can you please explain me what is the issue with the existing code 
swprintf_s(searchstr, MAX_PATH, L"%s\\*.*", wname);

**Review by bagajjal — CHANGES_REQUESTED** [review]

File: contrib/win32/win32compat/win32_dirent.c (hunk 14 above)
Comment (remittor) [reply contrib/win32/win32compat/win32_dirent.c:127]: Why need set zero?

File: contrib/win32/win32compat/misc.c (hunk 7 above)
Comment (remittor) [reply contrib/win32/win32compat/misc.c:204]: I think overwise.

File: contrib/win32/win32compat/win32_dirent.c (hunk 14 above)
Comment (bagajjal) [reply contrib/win32/win32compat/win32_dirent.c:127]: to avoid the corner cases like calling _findclose multiple times for the same hFile (or) trying to read on the hFile that is already closed.. setting the hFile to 0 inside the if loop will avoid them.

File: contrib/win32/win32compat/misc.c (hunk 7 above)
Comment (bagajjal) [reply contrib/win32/win32compat/misc.c:204]: minor comment, fix it as you like.

File: contrib/win32/win32compat/win32_dirent.c (hunk 14 above)
Comment (remittor) [reply contrib/win32/win32compat/win32_dirent.c:127]: You say it seriously?
After call `free(dirp);` read of field **hFile** cause an exception!
Content of field **hFile** located in heap memory!

Comment (bagajjal) [reply contrib/win32/win32compat/win32_dirent.c:127]: Yes, not required.. missed the free(dirp) while reviewing.

**Comment by bagajjal** [comment]
Hi remittor, Please update the latest code changes and submit the new review so that we can merge your changes.

**Comment by bagajjal** [comment]
@remittor - Could you please update pull request by updating your branch to latesw_all..

**Comment by bagajjal** [comment]
@remittor - Could you please update your branch to latestw_all and fix comments (if any)..
We would like to take this to may end release..

Few people from community are requesting for these changes..
https://github.com/PowerShell/Win32-OpenSSH/issues/539

**Comment by remittor** [comment]
@bagajjal I make a new PR today.
New **P**ull **R**equest: https://github.com/PowerShell/openssh-portable/pull/148
